### PR TITLE
Fail nicely when we can’t parse the rpc response

### PR DESCRIPTION
**Pull request by yosiat**
In case of invalid RPC response content (for example - invalid JSON message), we timeout and write error in the log “callback no more in memory” which is misguiding.

This change:
* Write “callback no more in memory” only if we fail to find the correlation id in the map
* Raise error to RPC caller in case we fail to parse the response
* Wraps the raised error in `BunnyError` so we can easily detect if we should retry - I want to do this for Timeout and Buffer Full but afraid to introduce compatibility issues.

### CODE REVIEW

**Comment by yosiat** [comment]
@mrister @Spring3 can someone look at this PR?

**Review by ykhatal — APPROVED** [review]

**Comment by ykhatal** [comment]
@yosiat Thanks for helping.
Sorry for the delay of the review, we are quite busy right now on other projects.

I will try to do a release when all reviews are done.

File: src/modules/producer.js [hunk 1]
```diff
@@ -8,6 +8,17 @@ const ERRORS = {
   BUFFER_FULL: 'Buffer is full'
 };
 
+class ProducerError extends Error {
+  constructor(error) {
+    super(error.message);
```
Comment (ofkindness) [inline src/modules/producer.js]: You can use `Destructuring assignment` here

File: src/modules/producer.js [hunk 2]
```diff
@@ -32,17 +43,27 @@ class Producer {
...
-      try {
-        // check the correlation ID sent by the initial message using RPC
-        const corrId = msg.properties.correlationId;
-        // if we found one, we execute the callback and delete it because it will never be received again anyway
-        rpcQueue[corrId].resolve(parsers.in(msg));
-        this._connection.config.transport.info('bmq:producer', `[${queue}] < answer`);
-        delete rpcQueue[corrId];
-      } catch (e) {
+      // check the correlation ID sent by the initial message using RPC
+      const corrId = msg.properties.correlationId;
```
Comment (ofkindness) [inline src/modules/producer.js]: And here

**Review by ofkindness — APPROVED** [review]

File: src/modules/producer.js (hunk 1 above)
Comment (yosiat) [reply src/modules/producer.js]: changed

File: src/modules/producer.js (hunk 2 above)
Comment (yosiat) [reply src/modules/producer.js]: changed

**Comment by yosiat** [comment]
hey, how can we expedite the process here?

**Review by Spring3 — APPROVED** [review]

**Review by mrister — APPROVED** [review]

**Comment by mrister** [comment]
@ky23 @Spring3 hope you guys can release this nice fix soon

**Pull request merged**